### ISSUE TITLE
Stack overflow error caused by serialization of `Map` with cyclic dependency -- NOT CVE

### ISSUE THREAD

**Issue by PoppingSnack** [FasterXML/jackson-databind#3972]
# Stack overflow error caused by jackson serialization Map

## Description

jackson before v2.15.2 was discovered to contain a stack overflow via the map parameter. This vulnerability allows attackers to cause a Denial of Service (DoS) via a crafted string.

## Error Log

```
Exception in thread "main" java.lang.StackOverflowError
	at java.base/java.lang.String.startsWith(String.java:1470)
	at com.fasterxml.jackson.databind.util.ClassUtil.isJDKClass(ClassUtil.java:1163)
	at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector._findStdTypeDesc(BasicClassIntrospector.java:258)
	at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(BasicClassIntrospector.java:80)
	at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(BasicClassIntrospector.java:11)
	at com.fasterxml.jackson.databind.SerializationConfig.introspect(SerializationConfig.java:906)
	at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.createKeySerializer(BasicSerializerFactory.java:210)
	at com.fasterxml.jackson.databind.SerializerProvider.findKeySerializer(SerializerProvider.java:915)
	at com.fasterxml.jackson.databind.SerializerProvider.findKeySerializer(SerializerProvider.java:926)
	at com.fasterxml.jackson.databind.ser.impl.PropertySerializerMap.findAndAddKeySerializer(PropertySerializerMap.java:144)
	at com.fasterxml.jackson.databind.ser.std.StdKeySerializers$Dynamic._findAndAddDynamic(StdKeySerializers.java:284)
	at com.fasterxml.jackson.databind.ser.std.StdKeySerializers$Dynamic.serialize(StdKeySerializers.java:262)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:797)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)



```

## PoC
```xml
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.15.2</version>
        </dependency>
```

```java
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.HashMap;

public class PoC2 {

    public static void main(String[] args) {
        HashMap<String,Object> map=new HashMap<>();
        map.put("t",map);
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            String jsonString = objectMapper.writeValueAsString(map);
            System.out.println(jsonString);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
    }
}
```

## Rectification Solution
1. Refer to the solution of jackson-databind: Add the depth variable to record the current parsing depth. If the parsing depth exceeds a certain threshold, an exception is thrown. (https://github.com/FasterXML/jackson-databind/commit/fcfc4998ec23f0b1f7f8a9521c2b317b6c25892b)

2. Refer to the GSON solution: Change the recursive processing on deeply nested arrays or JSON objects to stack+iteration processing.(（https://github.com/google/gson/commit/2d01d6a20f39881c692977564c1ea591d9f39027）)

## References

1. https://github.com/jettison-json/jettison/issues/52 
2. https://github.com/jettison-json/jettison/pull/53/files

**Comment by cowtowncoder** [comment]
Also quick note on Rectification Solutions: Jackson 2.15 does both -- default maximum nesting depth on reader side is set at 1000 levels, and `List` / `JsonNode` deserializers both have non-JDK-stack based handling.

**Comment by pjfanning** [comment]
@cowtowncoder I guess the pressure to make a change is going to come on - even if we do successfully engage with whoever assigned CVE-2023-35116 to get it revoked.

There are lots of ways to create data structures where you can get infinite recursion.
* Collections where one of the values is the top level Collection itself
* Classes with a parent type where the parent has already been seen. The simplest is to have an instance whose parent is itself. Can also be done with child types or sibling types.

The serialize APIs tend to take thse params - `(T value, JsonGenerator gen, SerializerProvider provider)`.

So as not to have to change the APIs to sneak in another value in the param list, could we consider adding something to SerializerProvider that allows us to track the state of the serialization? Specifically, we need to track the depth - how many levels down have we gone since the we started serializing the original instance? Alternatively, we could keep track of the values, so that if we are asked to serialize a child value, we can check if we have already serialized that value. I think depth is a safer approach - value equality is something that is hard to rely - some classes will have `equals()` implementations that will not suit our needs.

**Comment by cowtowncoder** [comment]
Whoever filed CVE https://www.mend.io/vulnerability-database/CVE-2023-35116 should go fuck themselves.

**Comment by ajakk** [comment]
> @PoppingSnack I assume you did not file this

I wouldn't be so sure...

![20230618_10h49m26s_grim](https://github.com/FasterXML/jackson-databind/assets/29188917/b600da61-aec8-46af-8a66-74bf130f0773)

**Comment by pjfanning** [comment]
@cowtowncoder @ajakk it most likely is that @PoppingSnack user - see https://github.com/janino-compiler/janino/issues/201 - another CVE raised despite the Janino team explaining that that tool is not designed for untrusted input - there has to be a point at which lib maintainers can say that users have to understand that they have to police against malicious input themselves

**Comment by yawkat** [comment]
This is laughable. Who's gonna file a CVE against JDK Map.hashCode which has the same "issue"?

**Comment by cowtowncoder** [comment]
Ok I gave too much benefit of doubt I guess. 

But fundamentally beyond blatant disregard to project accepting a report as vulnerability, reproduction here is not anything an attacker could do as shown. It is a standalone Foot Gun example in which developer of a service can DoS their own service. Ridiculous.
And description still makes baseless claim of "crafted String" -- while there is nothing of the sort in alleged reproduction.

So this is not in any way shape or from legit basis for a CVE: there is no attack vector for malicious clients.

**Comment by cowtowncoder** [comment]
@pjfanning Thank you for tackling that `jackson-core` issue that I filed earlier: I think that is the best way to go, based on similar thing on reader-side included in 2.15. There may still be reasons to tackle actual cyclic part (there have been PRs in the past proposing ways to keep track of parent object stack), but while related I think this one would tackle SO aspect reliably and efficiently.

(cycle tracking could actually possible be done at streaming level since there is hierarchic tracking of "current object" already... but that is still more complicated and adds significant overhead for deeper object hierarchies (so probably opt-in feature); whereas simple counter approach can be enabled for all users by default)

**Comment by AB-xdev** [comment]
For those coming from https://github.com/jeremylong/DependencyCheck:

So I just came in this morning and was a bit surprised to see that all our projects with the dependency - even with up to date versions - are marked as vulnerable.

The CVE is still tracked inside Sonatype's OSS INDEX (which is used by [DependencyCheck](https://github.com/jeremylong/DependencyCheck)) - I wrote them a mail to get this fixed.

**Comment by pjfanning** [comment]
@millems -- @cowtowncoder will be on vacation for a few weeks. My understanding is that Jackson 2.12 is no longer supported. Why not upgrade to a version that has the https://github.com/FasterXML/jackson-databind/commit/fcfc4998ec23f0b1f7f8a9521c2b317b6c25892b fix?

**Comment by millems** [comment]
Thanks for straightening me out. I've just been woken up by a page on [CVE-2023-35116](https://nvd.nist.gov/vuln/detail/CVE-2023-35116), so still a bit foggy. Our security policy requires us to treat literally every CVE as a problem. I can't push back on treating this as a security issue. I understand this issue isn't at all exploitable.

@pjfanning We've had a few conversations about this in the past with @cowtowncoder. It's a surprisingly large lift for us to upgrade to a new Jackson version if there have been any backwards-incompatible changes. We did it back in 2021: https://aws.amazon.com/blogs/developer/aws-sdk-for-java-version-1-12/. It sounds like we should do it again.

I have 2 short-term options:
1. Get CVE-2023-35116 revoked.
2. Appeal to the generosity of the Jackson developers to allow a backport of whatever "fix" was made for CVE-2023-35116 to 2.12.

Either way, I'll see about starting the process of getting Jackson upgraded to HEAD for all of our tools like we did back in 2021. I wasn't aware that 2.12 had officially moved to "not supported", although I am in no way surprised.

Which one of these options do you think is more likely to succeed?

**Comment by pjfanning** [comment]
@millems CVE-2023-35116 is not an issue. We reject that it is a security issue. There is nothing committed to jackson yet but we do intend to put some hardening in place in Jackson 2.16.0 - a release that is many months away. This fix will not be backported to any previous Jackson versions. https://github.com/FasterXML/jackson-core/pull/1055 is the proposed PR. To reiterate, this includes new APIs and will not be backported.

**Comment by millems** [comment]
Thanks! Is there anything we can do to help get CVE-2023-35116 revoked?

**Comment by yawkat** [comment]
@millems fyi the supported versions are listed here: https://github.com/FasterXML/jackson/wiki/Jackson-Releases – sounds like you should indeed upgrade.

(it incorrectly says 2.12 was closed in April 2022 when it was closed in April 2023, @pjfanning can you fix this? I don't have access.)

**Comment by pjfanning** [comment]
> @pjfanning Thanks! Is there anything we can do to help get [CVE-2023-35116](https://github.com/advisories/GHSA-gx6w-fqg7-mc3p) revoked?

I sent a request to Mitre using https://cveform.mitre.org/

I don't know if it is worth having multiple people chime in. I'll update this issue thread with any feedback from Mitre. There is no guarantee that they'll listen to me. A very high fraction of the CVEs are very low quality in terms of what they are reporting and often when they document real issues, the CVE description itself is still vague or inaccurate.

**Comment by millems** [comment]
Thanks! Sorry you have to deal with this. I'll tell our customers that you all are working to have the CVE revoked.

**Comment by DarekDan** [comment]
@PoppingSnack here is a cubic foot of stainless steel. Why don't you check if anything is vulnerable to dropping it from one yard of height in standard atmospheric conditions. Report back when ready with data to support your findings.

**Comment by yawkat** [comment]
@pjfanning have you heard back from MITRE? Otherwise i can try asking corp if they have anyone to contact there.

**Comment by pjfanning** [comment]
@yawkat I haven't heard back from MItre

**Comment by pjfanning** [comment]
the CVE has been updated to say 'DISPUTED' but they made the text worse by adding 'NOTE: the vendor's perspective is that the product is not intended for use with untrusted input.' 

This is absolutely not the case, this issue is about output not input. We have enhancements to police for malicious input.



https://github.com/CVEProject/cvelist/commit/e7481e1d5efe3ddea0cdb01706056b94f4bfb11e#diff-1d23b5f10a1be30eac5d043258c91f067a6f7e0cc86e7c6fc3e5c7462d6323f9

**Comment by jumarko** [comment]
Our dependency check jobs started failing again.
Is there any reasonable action other than to exclude this in the dep check configuration?

**Comment by AB-xdev** [comment]
> Our dependency check jobs started failing again.

https://nvd.nist.gov/vuln/detail/CVE-2023-35116#VulnChangeHistorySection

> Initial Analysis by NIST 6/26/2023 12:52:40 PM	
> Action | Type | Old Value | New Value
> -- | -- | -- | --
> Added | CPE Configuration |   | OR      ``*cpe:2.3:a:fasterxml:jackson-databind:*:*:*:*:*:*:*:*`` versions up to (including) 2.15.2
> Added | CVSS V3.1 |   | NIST AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H
> Added | CWE |   | NIST CWE-502
> Changed | Reference Type | https://github.com/FasterXML/jackson-databind/issues/3972 No Types Assigned | https://github.com/FasterXML/jackson-databind/issues/3972 Exploit, Issue Tracking


Congratulations to NIST looks like they are unable to read this issue properly and are also a week late. Maybe they should consider partnering with Internet Explorer.
Now they also classified it as affecting all Jackson Databind versions.... and according to them this issue contains an exploit... Yeah sure I also always copy java code into the system I want to exploit.

Also the CVSS score makes zero sense: https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?name=CVE-2023-35116&vector=AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H&version=3.1&source=NIST
* Attack vector = Network ?
* Attack complexity = Low ?

If someone has the motivation they may contact NIST and try to get it revoked. I'm not motivated anymore and going to file an issue at my company that we disable CVE scanning and just keep our dependencies up-to-date because this is a clown show.

**Comment by apetrelli** [comment]
Excuse me, the original reporter closed this issues on 19th June. What gives?

**Comment by apetrelli** [comment]
Excuse me, the original reporter closed this issues on 19th June. What gives?

**Comment by OrangeDog** [comment]
An inaccurate CVE in the database is OK, at least people can investigate and ignore it. But now it's lying about what "the vendor" said, which is much worse. How did that even happen? Just massive incompetence?

**Comment by cowtowncoder** [comment]
@OrangeDog I assume that org maintaining reports is bit like patent office; overworked and with too little resources to investigate in detail, mostly relying on reporters' competency and good will. And where last 2 are lacking results are like here. :-(

**Comment by Bhashkarjya** [comment]
I'm a bit late to the party.
So, I recently faced a security vulnerability for this issue in one of my Docker images. When can we expect for CVE-2023-35116 to get revoked?

**Comment by pjfanning** [comment]
> I'm a bit late to the party. So, I recently faced a security vulnerability for this issue in one of my Docker images. When can we expect for [CVE-2023-35116](https://github.com/advisories/GHSA-gx6w-fqg7-mc3p) to get revoked?

We have no control over the CVEs. Talk to whoever provides your security analysis tool.

**Comment by cowtowncoder** [comment]
Alas, project owners are at mercy of CVE maintainers, who can update state and descriptions.
It puzzles me, for example, how "project has the view that the library is not to be used for untrusted content" came about -- this is not the gist of what is being said here at all.

Rather, the project team considers this an invalid report: there is no valid reproduction of a security vulnerability.

**Comment by hdeweirdt** [comment]
I sent the NVD a message through their contact form, 🤞

**Comment by JDraganin** [comment]
Is it false positive?

**Comment by cowtowncoder** [comment]
> > The person who filed this CVE should really think
> 
> Closing the stable door after the horse has bolted.

"Why start now?" :)

**Comment by ajakk** [comment]
> I sent the NVD a message through their contact form, crossed_fingers

MITRE maintains the CVE data, not NVD. NVD is really just a frontend for that data. You need to talk to MITRE to get a CVE revoked or changed or anything:

https://cveform.mitre.org

**Comment by levinebw** [comment]
@cowtowncoder are you folks registered as CNAs for this project with MITRE? Not sure, but that might help get the issue removed.

**Comment by attritionorg** [comment]
It would. No other CNA could assign for your software in that case.

**Comment by cowtowncoder** [comment]
I don't think we are: I don't know what all that would entail. I did file CVEs via mitre UI when polymorphic deserialization issues were a thing (pre-2.10) and that was quite a bit of work, even without being an official contact point.

**Comment by pjfanning** [comment]
I'm not an expert but my understanding is that registering as a CNA for all com.fasterxml projects is pretty straightforward.

As a CNA, we could still register CVEs - probably via the cvemitre site like has already been done in the past. It would mean that Mitre should contact us if someone tries to register a CVE bypassing us. Reporters can still appeal to Mitre if we do not deal with the security report. I have seen some issues raised for ASF projects and that were not accepted as CVEs by the ASF being effectively appealed to Mitre. Mitre have then gotten back to the ASF to query the reasoning.

I would support going down the CNA route.

**Comment by cowtowncoder** [comment]
Ok. My only concern is with extra work for filing CVEs, but then again number is fairly low (after polymorphic deser flood), and at least we'd have more say in actually blocking invalid reports (and not just get optional FYI from submitter).
So I am not against application.

**Comment by hdeweirdt** [comment]
> > I sent the NVD a message through their contact form, crossed_fingers
> 
> MITRE maintains the CVE data, not NVD. NVD is really just a frontend for that data. You need to talk to MITRE to get a CVE revoked or changed or anything:
> 
> https://cveform.mitre.org

I sent a new report there.

**Comment by OrangeDog** [comment]
> We have also noted a reason for the dispute

Did you not point out that they reason they have noted is a lie?

**Comment by cowtowncoder** [comment]
I filed another request to at least try to get description changed to indicate that we actually disagree this is a valid vulnerability (reject).

**Comment by Rohit23110** [comment]
What is the latest update on this issue? Has MITRE caved in and indicated that this CVE would be removed?

**Comment by cowtowncoder** [comment]
I have not heard anything further for my submission via https://cveform.mitre.org/ . I assume anyone is free to file other requests, although we do not want to spam them and slow down processing.

**Comment by SavvasRuchiTengse** [comment]
Hello, how do I ignore this CVE in my dependency check in gradle ? I am new to gradle.

**Comment by vin01** [comment]
This was definitely filed by @PoppingSnack as shown on Snyk `Credit` for https://security.snyk.io/vuln/SNYK-JAVA-DEGROBMEIERJSON-5710359

Most likely it was reported via Snyk as well as they are also a CNA: https://snyk.io/blog/vulnerability-disclosure-program/

**Comment by Marcono1234** [comment]
As response to https://github.com/FasterXML/jackson-databind/issues/3972#issuecomment-1637273209:
> Ok. My only concern is with extra work for filing CVEs, but then again number is fairly low (after polymorphic deser flood), and at least we'd have more say in actually blocking invalid reports (and not just get optional FYI from submitter).

You can request CVE IDs through GitHub advisories (see [documentation](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/publishing-a-repository-security-advisory#requesting-a-cve-identification-number-optional)). Though I don't know if it would be possible (or allowed) that you are your own CNA, but let GitHub request the CVEs.

But since every maintainer of a GitHub project is now able to request CVE IDs on their own, maybe Mitre should just reject taking requests for GitHub projects, telling the reporter to contact the maintainer instead (and refer to the GitHub documentation, in case the maintainer is not aware that they can request CVE IDs easily)? And Mitre should only act then for arbitration requests, or when maintainers are unresponsive?

**Comment by cowtowncoder** [comment]
Yes, I have filed CVEs myself: that is not problematic.
It is also not the case that anyone SHOULD be able to file an issue without co-operating with maintainers -- it is the problem that while they SHOULD NOT, some still do. What happens, basically is that either:

1. Someone reports an alleged vulnerability and immediately files CVE ("we did notify maintainers!") but without waiting for any response
2. Someone just, well, files CVE without anything -- often based on a Github issue they think implicates a vulnerability.

So basically CVE filing is an honor system as things are (AFAIK). Becoming a CNA would, I think, not necessarily even prevent filing but give ability to modify CVEs to indicate they are valid. And possibly in some cases make Mitre even prevent the original filing.

**Comment by Marcono1234** [comment]
What I meant was, maybe we should try to convince Mitre (and any other CNAs such as Snyk) that they stop filing CVEs for GitHub repositories by default[^1][^2], and that issuing a CVE for a GitHub repo becomes a rare exceptional thing they do. So they only do it in cases of arbitration or for inactive maintainers; e.g. if a public GitHub issue mentioning a vulnerability without going into details and asking for contact details remains unanswered for a long time.

And that if someone approaches Mitre then and says "I would like a CVE for this vulnerability in GitHub repository _XYZ_", that they respond with something like this:
> Please contact the repository maintainer directly; check their "Security" page or contribution guide for contact details, or open an issue asking for contact details.
The maintainers of _XYZ_ can then request the CVE themselves. If they don't know how they can do that, refer them to the [GitHub documentation](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/publishing-a-repository-security-advisory#requesting-a-cve-identification-number-optional).

Also as side note, recently for curl a bogus CVE was issued as well; it appears they [consider becoming their own CNA now as well](https://daniel.haxx.se/blog/2023/09/05/bogus-cve-follow-ups/). But it is really not an ideal situation that maintainers have to become their own CNAs; the proper solution might be really if Mitre changed their behavior.

[^1]: Possibly except for cases where a GitHub repository is clearly only a mirror of a Git repository hosted somewhere else.
[^2]: Or of course unless those CNAs work together with the GitHub repo maintainers, for example when the maintainers are using HackerOne, which is also a CNA.

**Comment by cowtowncoder** [comment]
While I am all for changing behavior for better, I am somewhat cynical about likelihood of changing behavior of others.

Thank you for sharing your thoughts @Marcono1234 -- interesting thing about Curl, in particular.

**Comment by attritionorg** [comment]
Actually, becoming a CNA *would* prevent filing essentially. Then CNA rules tell every other CNA, including MITRE, to defer to you first for the assignment.

Please don't hold your breath for MITRE to change. There is no real incentive for them to do so.

**Comment by cowtowncoder** [comment]
@attritionorg Yes, I think this is the intent wrt becoming CNA. And I 100% agree in thinking it is rather unlikely anyone can easily change Mitre's SOP from outside -- my understanding is that there is widespread discontent wrt lax granting of CVEs, overblown Scary-Scoring and various other suboptimal aspects (f.ex Curl author made good points that I fully agree with).
But despite this being the case, I don't see any changes forthcoming -- and I also do not think they were unaware of this friction (i.e. politely telling them to stop blindly accepting garbage submission probably will do nothing).
Still, I am not going to stop anyone from trying to change things in positive way. Just suggest to tone down expectations.

**Comment by attritionorg** [comment]
@cowtowncoder I tried to change from the 'inside' when I was on the CVE Editorial Board for 10 years. They are well aware of many, many problems in the program. One problem is the board is full of a few that were actively trying to change things for the better, a lot of vendors that only cared about the program as how it affected them, and a ton of lurkers. Since then, the program has gotten dramatically worse in many ways.

**Comment by cowtowncoder** [comment]
@attritionorg Alas, this is very close to what I was guessing might be the case. :-(
I wish it wasn't.

Thank you for sharing -- I do not have internal view, just external observations.

**Comment by yawkat** [comment]
Maybe it would be worth it to join forces with the curl project (and/or other large OSS projects) to become a CNA.

**Comment by attritionorg** [comment]
@yawkat Unless there is some 'official' organization that encompasses many projects, that likely isn't possible and doesn't fall under the scope of an 'organization' or business that qualifies to become a CNA. Especially with GitHub being a CNA and every project rolling up to them for assignments if any given repo/project opts to. It's a pain in the ass I bet, but any high-profile project or library should look to become one in my opinion, ONLY if they are seeing a lot of bad assignments. Then it allows you take the power away from MITRE or other CNAs on assignment. At worst, a researcher can protest to MITRE that their vuln should get one, but then MITRE would defer to you... and more importantly, a conversation would happen between you and MITRE. If it didn't, then it means MITRE is further breaking down their own polices and procedures, which have warped considerably the last 5 - 10 years.

**Comment by Marcono1234** [comment]
I suggested in https://github.com/github/advisory-database/issues/2718 to have GitHub extend its CNA scope, possibly similar to what GitLab is doing for their platform. Maybe that could improve the situation as well, but let's see what the GitHub employees think about this.

**Comment by cowtowncoder** [comment]
@Marcono1234 Ah. Did not know Gitlab is doing this (I used them at my previous company, they seemed to have a few things that were ahead of Github, but are much less well known). Thank you for filing that issue (regardless of whether anything happens -- at least it can be pointed to as one possible way forward, plus exposes the issues).

**Comment by tonyschwartz** [comment]
I do believe this needs to be fixed.  Any REST application receives json data from potentially malicious users.  There is no mechanism to test the validity of that json data outside the context of the json parser.  It is the responsibility of the parser to detect malicious json data.  The example provided above by @yawkat with a.next and b.next pointing to each other is not the same thing.  That would be the developer's fault, not the fault of a malicious actor.  If it is possible for a malicious actor to craft a data structure that when supplied to a RESTful service, crashes that service, then there is a security vulnerability.  If the json parsing framework is not capable of detecting this without stack overflow and crashing, then the json parsing framework cannot be trusted.  This is my two cents and not meant to be rude or offensive to anyone.  Please correct me where I'm wrong here.

**Comment by yawkat** [comment]
@tonyschwartz 

> If it is possible for a malicious actor to craft a data structure that when supplied to a RESTful service, crashes that service, then there is a security vulnerability

It is not possible to construct such a data structure from json.

**Comment by cowtowncoder** [comment]
@tonyschwartz Please re-read the discussion again with thought. THERE IS NO VULNERABILITY.
As @yawkat suggests JSON does not allow expressing of cyclic data -- nor does Jackson object mapper by default.

So given PoC for assumed vulnerability is specifically constructed to allow cycles and then specifically crafted JSON that PoC is sending produces SOE. That is not vulnerability: attacker has way to force use of specific data model.

**Comment by tonyschwartz** [comment]
I believe it is possible though.  The stack is only so big.  If I send you a 20MB json with many child objects, I can crash your app, no?
{"a":{"a":{...

**Comment by cowtowncoder** [comment]
@tonyschwartz That is different issue, solved via #943 (for Jackson 2.15.0).

**Comment by yawkat** [comment]
@tonyschwartz we have protection against deeply nested structures

**Comment by tonyschwartz** [comment]
I see.  So, I'm a bit confused on the CVE then.  I will study.

**Comment by tonyschwartz** [comment]
I hear you and agree completely.  But, the authors of the CVE say, "allows attackers to cause a denial of service or other unspecified impact via a crafted object".  It implies the attacker has control, not the developer.  I trust you have ruled this out and my comment was not intended to refute your claim.  But, the CVE, as written, does imply there is more to the story.  I would like to talk to the person/team who authored the CVE.  Thanks for the reply!

**Comment by yawkat** [comment]
yes, we've discussed extensively that the CVE is wrong, but apparently we can't do anything about it.

**Comment by cowtowncoder** [comment]
@tonyschwartz That statement made by submitted of CVE is indeed factually incorrect. I have tried to make Mitre change the description as have others -- to no avail. In fact, the description became even worse (but I digress).
Mitre folks do not really seem to pay much attention (due to lack of resources etc) to correctness of reports; I assume they think that reporters and maintainers should somehow sort it all out without their involvement.
The whole system is rotten to its core.

I suggest you read, say https://daniel.haxx.se/blog/2023/09/05/bogus-cve-follow-ups/ (if you haven't) -- it sounds like you may overestimating integrity (and perhaps competence) of people submitting CVEs.
I do not want to rehash everything said there but agree 100%.

Literally anyone, anywhere, can file bogus CVEs and have a reasonable change of that being published and causing lots of FUD without being based on anything solid.

**Comment by cowtowncoder** [comment]
Some minor positive update: NIST has upgraded description at https://nvd.nist.gov/vuln/detail/CVE-2023-35116 to include:

   NOTE: the vendor's perspective is that this is not a valid vulnerability report, because the steps of constructing a cyclic data structure and trying to serialize it cannot be achieved by an external attacker

which is accurate. We do not consider this a valid vulnerability.

Having said that, SHOULD there be a way to add cyclic dependency, serialization would be blocked by:

https://github.com/FasterXML/jackson-core/pull/1055

to be included in Jackson 2.16.

**Comment by kashok7474** [comment]
which version will address https://nvd.nist.gov/vuln/detail/CVE-2023-35116?

**Comment by pjfanning** [comment]
> which version will address https://nvd.nist.gov/vuln/detail/CVE-2023-35116?

This has been answered in https://github.com/FasterXML/jackson-databind/issues/3972#issuecomment-1742344745 - 2.16.0-rc1 is released already and 2.16.0 should be out soon.

**Comment by cowtowncoder** [comment]
To prevent further spamming by folks asking for already answered question -- CVE-2023-35116 is INVALID BASELESS BOGUS NO GOO FALSE ALARM -- will try closing comments to this issue.
